commands/builderremote_linux.go

```diff
@@ -0,0 +1,396 @@
+//go:build linux
+
+package commands
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"os/exec"
+	"os/signal"
+	"path/filepath"
+	"syscall"
+	"time"
+
+	"github.com/containerd/containerd/log"
+	"github.com/docker/buildx/build"
+	"github.com/docker/buildx/commands/builder"
+	builderapi "github.com/docker/buildx/commands/builder/pb"
+	"github.com/docker/buildx/monitor"
+	"github.com/docker/buildx/version"
+	"github.com/docker/cli/cli/command"
+	"github.com/moby/buildkit/client"
+	"github.com/pelletier/go-toml"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"google.golang.org/grpc"
+)
+
+const (
+	launchCommandName = "_INTERNAL_LAUNCH"
+	serveCommandName  = "_INTERNAL_SERVE"
+)
+
+type serverConfig struct {
+	// Specify buildx server root
+	Root string `toml:"root"`
+
+	// LogLevel sets the logging level [trace, debug, info, warn, error, fatal, panic]
+	LogLevel string `toml:"log_level"`
+
+	// Specify file to output buildx server log
+	LogFile string `toml:"log_file"`
+}
+
+func newRemoteBuildxController(ctx context.Context, opts buildOptions) (monitor.BuildxController, error) {
+	rootDir := opts.root
+	if rootDir == "" {
+		var err error
+		rootDir, err = rootDataDir()
+		if err != nil {
+			return nil, err
+		}
+	}
+	serverRoot := filepath.Join(rootDir, "shared")
+	c, err := newBuildxClientAndCheck(filepath.Join(serverRoot, "buildx.sock"), 1, 0)
+	if err != nil {
+		logrus.Info("no buildx server found; launching...")
+		// start buildx server via subcommand
+		launchFlags := []string{}
+		if opts.serverConfig != "" {
+			launchFlags = append(launchFlags, "--config", opts.serverConfig)
+		}
+		logFile, err := getLogFilePath(opts.serverConfig)
+		if err != nil {
+			return nil, err
+		}
+		wait, err := launch(ctx, true, logFile, append([]string{launchCommandName}, launchFlags...)...)
+		if err != nil {
+			return nil, err
+		}
+		if err := wait(); err != nil {
+			return nil, err
+		}
+		c, err = newBuildxClientAndCheck(filepath.Join(serverRoot, "buildx.sock"), 10, time.Second)
+		if err != nil {
+			return nil, fmt.Errorf("cannot connect to the buildx server: %w", err)
+		}
+	}
+	return &buildxController{c, serverRoot}, nil
+}
+
+func addBuilderCommands(cmd *cobra.Command, dockerCli command.Cli, rootOpts *rootOptions) {
+	cmd.AddCommand(
+		launchCmd(dockerCli, rootOpts),
+		serveCmd(dockerCli, rootOpts),
+	)
+}
+
+func launchCmd(dockerCli command.Cli, rootOpts *rootOptions) *cobra.Command {
```

Comment:
Why do we need both `launchCmd` and `serveCmd`? It seems like `launch` just calls `serve`, but I might be missing something as to why it's necessary.

Reply:
This was for immeditely detaching buildx server processs from the buildx (client) process and reparenting it to the init process. But it's not strictly necessary because reparenting happens after buildx (client) exits.
Fixed to directly exec buildx server.